File: backend/tools/sdk-ts/draox-client/src/DraoxClient.ts
import { EventEmitter } from 'events';
import { WebSocketTransport } from './transports/WebSocketTransport';
import type { ITransport } from './transports/ITransport';
import type { DraoxConfig, DraoxEvent, ResolvedConfig, ClientState, LoginResponse } from './types';
import { Serializer } from './Serializer';
import { RequestBroker } from './RequestBroker';
import { Reconnector } from './Reconnector';

export declare interface DraoxClient {
  on(event: 'connected',    listener: () => void): this;
  on(event: 'disconnected', listener: (reason: string) => void): this;
  on(event: 'authenticated',listener: () => void): this;
  on(event: 'stateChanged', listener: (state: ClientState) => void): this;
  on(event: 'error',        listener: (err: Error) => void): this;
}

export class DraoxClient extends EventEmitter {
  private readonly cfg: ResolvedConfig;
  private transport:    ITransport | null = null;
  private broker:       RequestBroker;
  private reconnector:  Reconnector;
  private heartbeat:    ReturnType<typeof setInterval> | null = null;
  private abort:        AbortController | null = null;
  private missedPings = 0;

  private savedUserId: string | null = null;
  private savedToken:  string | null = null;

  private _state:     ClientState = 'disconnected';
  private _sessionId: string | null = null;

  private readonly eventHandlers    = new Map<string, Set<(e: DraoxEvent) => void>>();
  private readonly categoryHandlers = new Map<string, Set<(e: DraoxEvent) => void>>();

  get state():           ClientState    { return this._state; }
  get sessionId():       string | null  { return this._sessionId; }
  get isAuthenticated(): boolean        { return this._sessionId !== null; }

  constructor(config: DraoxConfig = {}) {
    super();
    const protocol = config.protocol ?? 'ws';
    const defaultPort = protocol === 'grpc' ? 9004 : 9002;
    this.cfg = {
      host:                config.host                ?? 'localhost',
      port:                config.port                ?? defaultPort,
      adminPort:           config.adminPort           ?? 9100,
      protocol,
      useTls:              config.useTls              ?? false,
      timeoutMs:           config.timeoutMs           ?? 10_000,
      heartbeatIntervalMs: config.heartbeatIntervalMs ?? 30_000,
      reconnect: {
        enabled:     config.reconnect?.enabled     ?? true,
        maxAttempts: config.reconnect?.maxAttempts ?? 5,
        baseDelayMs: config.reconnect?.baseDelayMs ?? 1_000,
        maxDelayMs:  config.reconnect?.maxDelayMs  ?? 30_000,
      },
      grpc: {
        protoPath:   config.grpc?.protoPath   ?? '',
        credentials: config.grpc?.credentials ?? 'insecure',
      },
    };
    this.broker      = new RequestBroker();
    this.reconnector = new Reconnector(this.cfg.reconnect);
  }

  // ── Public API ────────────────────────────────────────────────────────────

  async connect(): Promise<void> {
    if (this._state === 'connected' || this._state === 'connecting') return;

    this.abort = new AbortController();
    this.setState('connecting');

    this.transport = new WebSocketTransport();
    this.transport.onMessage = (msg) => this.onMessage(msg);
    this.transport.onClose   = (reason) => this.onClosed(reason);

    await this.transport.connect(this.cfg.host, this.cfg.port, this.cfg.useTls);

    this.setState('connected');
    this.emit('connected');
    this.startHeartbeat();
  }

  async disconnect(reason = 'client_disconnect'): Promise<void> {
    this.stopHeartbeat();
    this.abort?.abort();

    if (this.transport) {
      this.broker.failAll(new Error(`Disconnected: ${reason}`));
      this.transport.disconnect();
      this.transport = null;
    }

    this._sessionId = null;
    this.setState('disconnected');
    this.emit('disconnected', reason);
  }

  async authenticate(userId: string, token: string): Promise<void> {
    this.savedUserId = userId;
    this.savedToken  = token;

    const data = await this.requestInternal<{ session_id: string }>('auth', { user_id: userId, token });
    this._sessionId = data.session_id;
    this.emit('authenticated');
  }

  async login(username: string, password: string): Promise<void> {
    const scheme = this.cfg.useTls ? 'https' : 'http';
    const url    = `${scheme}://${this.cfg.host}:${this.cfg.adminPort}/api/auth/login`;

    const res = await fetch(url, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify({ username, password }),
    });

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Login failed (${res.status}): ${text}`);
    }

    const body = await res.json() as { success: boolean; data: LoginResponse };
    if (!body.success || !body.data?.token)
      throw new Error('Login failed: unexpected response format');

    await this.authenticate(body.data.username, body.data.token);
  }

  async send(action: string, payload?: unknown): Promise<void> {
    this.ensureConnected();
    const json = Serializer.serialize({ id: this.newId(), type: 'request', action, payload });
    await this.transport!.send(json);
  }

  async request<T>(action: string, payload?: unknown): Promise<T> {
    this.ensureConnected();
    return this.requestInternal<T>(action, payload);
  }

  subscribe(eventName: string, handler: (e: DraoxEvent) => void): void {
    this.addHandler(this.eventHandlers, eventName, handler);
  }

  unsubscribe(eventName: string, handler: (e: DraoxEvent) => void): void {
    this.removeHandler(this.eventHandlers, eventName, handler);
  }

  subscribeCategory(category: string, handler: (e: DraoxEvent) => void): void {
    this.addHandler(this.categoryHandlers, category, handler);
  }

  unsubscribeCategory(category: string, handler: (e: DraoxEvent) => void): void {
    this.removeHandler(this.categoryHandlers, category, handler);
  }

  // ── Internal ──────────────────────────────────────────────────────────────

  private async requestInternal<T>(action: string, payload?: unknown): Promise<T> {
    const id   = this.newId();
    const json = Serializer.serialize({ id, type: 'request', action, payload });
    const res  = await this.broker.send(
      (j) => this.transport!.send(j), json, id, this.cfg.timeoutMs,
    );
    if (!res.success) throw new Error(res.error ?? 'request failed');
    return res.data as T;
  }

  private onMessage(json: string): void {
    const msg = Serializer.parse(json);
    if (!msg) return;

    switch (msg.type) {
      case 'response':
        if (msg.id)
          this.broker.complete(msg.id, {
            id:      msg.id,
            success: msg.success ?? false,
            data:    msg.data,
            error:   msg.error,
          });
        break;

      case 'event':
        if (msg.category && msg.name)
          this.dispatchEvent({
            category:  msg.category,
            name:      msg.name,
            data:      msg.data,
            timestamp: msg.timestamp ?? new Date().toISOString(),
          });
        break;

      case 'pong':
        this.missedPings = 0;
        break;
    }
  }

  private dispatchEvent(evt: DraoxEvent): void {
    this.eventHandlers.get(`${evt.category}.${evt.name}`)?.forEach(h => h(evt));
    this.eventHandlers.get(evt.name)?.forEach(h => h(evt));
    this.categoryHandlers.get(evt.category)?.forEach(h => h(evt));
  }

  private onClosed(reason: string): void {
    this.stopHeartbeat();
    this.setState('disconnected');
    this.emit('disconnected', reason);

    if (this.cfg.reconnect.enabled && this.abort && !this.abort.signal.aborted)
      void this.tryReconnect();
  }

  private async tryReconnect(): Promise<void> {
    this.setState('reconnecting');

    const success = await this.reconnector.attempt(async () => {
      try {
        await this.transport!.connect(this.cfg.host, this.cfg.port, this.cfg.useTls);
        if (this.savedUserId) await this.authenticate(this.savedUserId, this.savedToken!);
        return true;
      } catch { return false; }
    }, this.abort!.signal);

    if (success) {
      this.setState('connected');
      this.emit('connected');
      this.startHeartbeat();
    }
  }

  private startHeartbeat(): void {
    this.stopHeartbeat();
    this.heartbeat = setInterval(async () => {
      if (!this.transport?.isConnected) { this.stopHeartbeat(); return; }
      this.missedPings++;
      if (this.missedPings >= 2) { this.onClosed('heartbeat_timeout'); return; }
      try {
        await this.transport.send(Serializer.serialize({ type: 'ping', ts: Date.now() }));
      } catch { this.stopHeartbeat(); }
    }, this.cfg.heartbeatIntervalMs);
  }

  private stopHeartbeat(): void {
    if (this.heartbeat) { clearInterval(this.heartbeat); this.heartbeat = null; }
  }

  private ensureConnected(): void {
    if (this._state !== 'connected')
      throw new Error(`Not connected (state: ${this._state})`);
  }

  private addHandler(map: Map<string, Set<(e: DraoxEvent) => void>>, key: string, h: (e: DraoxEvent) => void): void {
    if (!map.has(key)) map.set(key, new Set());
    map.get(key)!.add(h);
  }

  private removeHandler(map: Map<string, Set<(e: DraoxEvent) => void>>, key: string, h: (e: DraoxEvent) => void): void {
    map.get(key)?.delete(h);
  }

  private newId(): string {
    return `req_${crypto.randomUUID().replace(/-/g, '')}`;
  }

  private setState(state: ClientState): void {
    if (this._state === state) return;
    this._state = state;
    this.emit('stateChanged', state);
  }
}

File: backend/tools/sdk-wpf/DraoxClientWpf/Core/DraoxClient.cs
using System.Net.Http;
using System.Text;
using System.Text.Json;

namespace Draox.Client;

/// <summary>
/// Draox client SDK for .NET desktop applications.
/// Create an instance, call ConnectAsync(), then AuthenticateAsync() before using the messaging APIs.
/// </summary>
public class DraoxClient : IDisposable
{
    private readonly DraoxConfig _config;
    private IConnection? _connection;
    private RequestBroker? _broker;
    private Reconnector? _reconnector;
    private CancellationTokenSource? _lifetimeCts;
    private CancellationTokenSource? _heartbeatCts;
    private readonly SynchronizationContext? _syncCtx;

    private string? _savedUserId;
    private string? _savedToken;
    private string? _savedUsername;
    private string? _savedPassword;
    private int _missedPings;

    private static readonly HttpClient _http = new();

    private readonly Dictionary<string, List<Action<DraoxEvent>>> _eventHandlers    = new();
    private readonly Dictionary<string, List<Action<DraoxEvent>>> _categoryHandlers = new();

    public ClientState State     { get; private set; } = ClientState.Disconnected;
    public string?     SessionId { get; private set; }
    public bool        IsAuthenticated => !string.IsNullOrEmpty(SessionId);
    public DraoxConfig Config    => _config;

    public event Action?              OnConnected;
    public event Action<string>?      OnDisconnected;
    public event Action<string>?      OnError;
    public event Action?              OnAuthenticated;
    public event Action<ClientState>? OnStateChanged;

    public DraoxClient(DraoxConfig? config = null)
    {
        _config  = config ?? new DraoxConfig();
        _syncCtx = SynchronizationContext.Current;
    }

    // ── Public API ────────────────────────────────────────────────────────────

    public async Task ConnectAsync(CancellationToken ct = default)
    {
        if (State is ClientState.Connected or ClientState.Connecting) return;

        _lifetimeCts = new CancellationTokenSource();
        SetState(ClientState.Connecting);

        _connection = _config.Protocol switch
        {
            DraoxProtocol.Tcp => new TcpConnection(),
            _                 => new WebSocketConnection(),
        };

        _broker      = new RequestBroker();
        _reconnector = new Reconnector(_config.Reconnect);

        _connection.MessageReceived += OnMessageReceived;
        _connection.Closed          += OnConnectionClosed;

        await _connection.ConnectAsync(_config, ct);

        SetState(ClientState.Connected);
        RaiseEvent(() => OnConnected?.Invoke());
        StartHeartbeat();
    }

    public async Task DisconnectAsync(string reason = "client_disconnect")
    {
        _heartbeatCts?.Cancel();
        _lifetimeCts?.Cancel();

        if (_connection is not null)
        {
            _broker?.FailAll(new DraoxException($"Disconnected: {reason}"));
            await _connection.DisconnectAsync();
            _connection = null;
        }

        SessionId = null;
        SetState(ClientState.Disconnected);
        RaiseEvent(() => OnDisconnected?.Invoke(reason));
    }

    public async Task AuthenticateAsync(string userId, string token, CancellationToken ct = default)
    {
        _savedUserId = userId;
        _savedToken  = token;

        var data = await RequestInternalAsync<AuthResponseData>(
            "auth", new { user_id = userId, token }, ct);
        SessionId = data?.SessionId;
        RaiseEvent(() => OnAuthenticated?.Invoke());
    }

    public async Task LoginAsync(string username, string password, CancellationToken ct = default)
    {
        _savedUsername = username;
        _savedPassword = password;

        var scheme = _config.UseTls ? "https" : "http";
        var url    = $"{scheme}://{_config.Host}:{_config.AdminPort}/api/auth/login";
        var body   = JsonSerializer.Serialize(new { username, password });
        using var req = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json"),
        };
        using var resp = await _http.SendAsync(req, ct);
        var json = await resp.Content.ReadAsStringAsync(ct);

        if (!resp.IsSuccessStatusCode)
            throw new DraoxException($"Login failed ({(int)resp.StatusCode}): {json}");

        using var doc  = JsonDocument.Parse(json);
        var root = doc.RootElement;
        if (!root.TryGetProperty("success", out var ok) || !ok.GetBoolean())
            throw new DraoxException("Login failed: server returned success=false");

        var data     = root.GetProperty("data");
        var token    = data.GetProperty("token").GetString()
                       ?? throw new DraoxException("Login failed: missing token");
        var user     = data.TryGetProperty("username", out var u) ? u.GetString() ?? username : username;

        await AuthenticateAsync(user, token, ct);
    }

    // Fire-and-forget.
    public Task SendAsync(string action, object? payload = null, CancellationToken ct = default)
    {
        EnsureConnected();
        var json = Serializer.Serialize(
            new WireRequest { Id = NewId(), Action = action, Payload = Serializer.ToNode(payload) });
        return _connection!.SendTextAsync(json, ct);
    }

    // Awaited request/response.
    public async Task<T?> RequestAsync<T>(string action, object? payload = null, CancellationToken ct = default)
    {
        EnsureConnected();
        return await RequestInternalAsync<T>(action, payload, ct);
    }

    public void Subscribe(string eventName, Action<DraoxEvent> handler)
        => AddHandler(_eventHandlers, eventName, handler);

    public void Unsubscribe(string eventName, Action<DraoxEvent> handler)
        => RemoveHandler(_eventHandlers, eventName, handler);

    public void SubscribeCategory(string category, Action<DraoxEvent> handler)
        => AddHandler(_categoryHandlers, category, handler);

    public void UnsubscribeCategory(string category, Action<DraoxEvent> handler)
        => RemoveHandler(_categoryHandlers, category, handler);

    public void Dispose() => DisconnectAsync().GetAwaiter().GetResult();

    // ── Internal ──────────────────────────────────────────────────────────────

    private async Task<T?> RequestInternalAsync<T>(string action, object? payload, CancellationToken ct)
    {
        var id   = NewId();
        var json = Serializer.Serialize(
            new WireRequest { Id = id, Action = action, Payload = Serializer.ToNode(payload) });
        var res  = await _broker!.SendAsync(_connection!, json, id, _config.TimeoutMs, ct);
        if (!res.Success) throw new DraoxException(res.Error ?? "request failed");
        return Serializer.Deserialize<T>(res.RawData);
    }

    private void OnMessageReceived(string json)
    {
        var msg = Serializer.Parse(json);
        if (msg is null) return;

        switch (msg.Type)
        {
            case "response":
                _broker?.Complete(msg.Id!, new DraoxResponse
                {
                    Id      = msg.Id!,
                    Success = msg.Success,
                    RawData = msg.RawData,
                    Error   = msg.Error,
                });
                break;

            case "event":
                DispatchEvent(new DraoxEvent
                {
                    Category  = msg.Category ?? "",
                    Name      = msg.Name ?? "",
                    RawData   = msg.RawData,
                    Timestamp = msg.Timestamp,
                });
                break;

            case "pong":
                _missedPings = 0;
                break;
        }
    }

    private void DispatchEvent(DraoxEvent evt)
    {
        TryDispatch(_eventHandlers,    $"{evt.Category}.{evt.Name}", evt);
        TryDispatch(_eventHandlers,    evt.Name,                     evt);
        TryDispatch(_categoryHandlers, evt.Category,                 evt);
    }

    private void TryDispatch(Dictionary<string, List<Action<DraoxEvent>>> dict, string key, DraoxEvent evt)
    {
        if (dict.TryGetValue(key, out var handlers))
            foreach (var h in handlers)
                RaiseEvent(() => h(evt));
    }

    private void OnConnectionClosed(string reason)
    {
        _heartbeatCts?.Cancel();
        SetState(ClientState.Disconnected);
        RaiseEvent(() => OnDisconnected?.Invoke(reason));

        if (_config.Reconnect.Enabled && _reconnector is not null
            && _lifetimeCts is { IsCancellationRequested: false })
            _ = TryReconnectAsync();
    }

    private async Task TryReconnectAsync()
    {
        SetState(ClientState.Reconnecting);

        var success = await _reconnector!.AttemptAsync(async () =>
        {
            try
            {
                await _connection!.ConnectAsync(_config, _lifetimeCts!.Token);
                if (_savedUsername is not null && _savedPassword is not null)
                    await LoginAsync(_savedUsername, _savedPassword, _lifetimeCts.Token);
                else if (!string.IsNullOrEmpty(_savedUserId))
                    await AuthenticateAsync(_savedUserId, _savedToken!, _lifetimeCts.Token);
                return true;
            }
            catch { return false; }
        }, _lifetimeCts!.Token);

        if (success)
        {
            SetState(ClientState.Connected);
            RaiseEvent(() => OnConnected?.Invoke());
            StartHeartbeat();
        }
    }

    private void StartHeartbeat()
    {
        _heartbeatCts?.Cancel();
        _heartbeatCts = new CancellationTokenSource();
        _ = HeartbeatLoopAsync(_heartbeatCts.Token);
    }

    private async Task HeartbeatLoopAsync(CancellationToken ct)
    {
        var interval = TimeSpan.FromSeconds(_config.HeartbeatIntervalSeconds);
        while (!ct.IsCancellationRequested)
        {
            try { await Task.Delay(interval, ct); }
            catch (OperationCanceledException) { break; }

            if (_connection is null || !_connection.IsConnected) break;

            _missedPings++;
            if (_missedPings >= 2) { OnConnectionClosed("heartbeat_timeout"); break; }

            var ping = Serializer.Serialize(new PingMessage { Ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() });
            try { await _connection.SendTextAsync(ping, ct); }
            catch { break; }
        }
    }

    private void EnsureConnected()
    {
        if (State != ClientState.Connected)
            throw new DraoxException($"Not connected (state: {State})");
    }

    private void RaiseEvent(Action action)
    {
        if (_syncCtx is not null && SynchronizationContext.Current != _syncCtx)
            _syncCtx.Post(_ => action(), null);
        else
            action();
    }

    private static void AddHandler(Dictionary<string, List<Action<DraoxEvent>>> dict, string key, Action<DraoxEvent> h)
    {
        if (!dict.TryGetValue(key, out var list)) dict[key] = list = new();
        list.Add(h);
    }

    private static void RemoveHandler(Dictionary<string, List<Action<DraoxEvent>>> dict, string key, Action<DraoxEvent> h)
    {
        if (dict.TryGetValue(key, out var list)) list.Remove(h);
    }

    private static string NewId() => $"req_{Guid.NewGuid():N}";

    private void SetState(ClientState state)
    {
        if (State == state) return;
        State = state;
        RaiseEvent(() => OnStateChanged?.Invoke(state));
    }
}
